File: CPP/SudokuSolver.cpp
class Solution {
public:
    
    bool isValid(vector<vector<char>>& board, int r, int c, char ch){
        
        for(int i = 0; i < 9; i++){
            if(board[i][c] == ch) return false;
            
            if(board[r][i] == ch) return false;
            
            if(board[3*(r/3)+i/3][3*(c/3)+i%3] == ch) return false;
        }
        return true;
    }
    
    bool BT(vector<vector<char>>& board){
        for(int i = 0; i < 9; i++){
            for(int j = 0; j < 9; j++){
                if(board[i][j] == '.'){
                    for(char c = '1'; c <= '9'; c++){
                        if(isValid(board, i, j, c)){
                            board[i][j] = c;
                            if(BT(board)) return true;
                            else{
                                board[i][j] = '.';
                            }
                        }
                            
                        
                    }
                    return false;
                }
                
            }
        }
        return true;
    }
    
    
    void solveSudoku(vector<vector<char>>& board) {
        
        BT(board);
        
    }
};

File: DSA/CSES_ProblemSet/SudokuSolver.java
class Solution {
    public void solveSudoku(char[][] board) {
        backtrack(board, 0, 0);
    }

        boolean backtrack(char[][] board, int i, int j) {
        int m = 9, n = 9;
        if (j == n) {
            
            return backtrack(board, i + 1, 0);
        }
        if (i == m) {
            
            return true;
        }

        if (board[i][j] != '.') {
            
            return backtrack(board, i, j + 1);
        } 

        for (char ch = '1'; ch <= '9'; ch++) {
            
            if (!isValid(board, i, j, ch))
                continue;
            
            board[i][j] = ch;
            
            if (backtrack(board, i, j + 1)) {
                return true;
            }
            board[i][j] = '.';
        }
        
        return false;
    }

    boolean isValid(char[][] board, int r, int c, char n) {
        for (int i = 0; i < 9; i++) {
            if (board[r][i] == n) return false;
            if (board[i][c] == n) return false;
            if (board[(r/3)*3 + i/3][(c/3)*3 + i%3] == n)
                return false;
        }
        return true;
    }
}
